Mathlib/Tactic/MinImports.lean

```diff
@@ -37,20 +37,65 @@ def getVisited (env : Environment) (decl : Name) : NameSet :=
 Otherwise it returns `stx` itself. -/
-def getId (stx : Syntax) : Syntax :=
+def getId (stx : Syntax) : CommandElabM Syntax := do
+  -- if the command contains a `declId`, we use the implied `ident`
   match stx.find? (·.isOfKind ``Lean.Parser.Command.declId) with
-    | some declId => declId[0]
-    | none => stx
-
-/--`getAllImports cmd` takes a `Syntax` input `cmd` and returns the `NameSet` of all the
-module names that are implied by the `SyntaxNodeKinds` and the identifiers contained in `cmd`. -/
-def getAllImports {m : Type → Type} [Monad m] [MonadResolveName m] [MonadEnv m]
-    (cmd : Syntax) (dbg? : Bool := false) :
-    m NameSet := do
+    | some declId => return declId[0]
+    | none =>
+      -- otherwise, there is the possibility that the command is a nameless `instance`
+      match stx.find? (·.isOfKind ``Lean.Parser.Command.instance) with
+                | none => return stx
+                | some stx => do
+                  -- if it is a nameless `instance`, we retrieve the autogenerated name
+                  let dv ← mkDefViewOfInstance {} stx
+                  return dv.declId[0]
+
+/-- `getIds stx` extracts all identifiers, collecting them in a `NameSet`. -/
+partial
+def getIds : Syntax → NameSet
+  | .node _ _ args => (args.map getIds).foldl (·.append ·) {}
+  | .ident _ _ nm _ => NameSet.empty.insert nm
+  | _ => {}
+
+/-- `getAttrNames stx` extracts `attribute`s from `stx`.
+It does not collect `simp`, `ext`, `to_additive`.
+It should collect almost all other attributes, e.g., `fun_prop`. -/
+def getAttrNames (stx : Syntax) : NameSet :=
+  match stx.find? (·.isOfKind ``Lean.Parser.Term.attributes) with
+    | none => {}
+    | some stx => getIds stx
+
+/-- `getAttrs env stx` returns all attribute declaration names contained in `stx` and registered
+in the `Environment `env`. -/
+def getAttrs (env : Environment) (stx : Syntax) : NameSet :=
+  Id.run do
+  let mut new : NameSet := {}
+  for attr in (getAttrNames stx) do
+    match getAttributeImpl env attr with
+      | .ok attr => new := new.insert attr.ref
+      | .error .. => pure ()
+  return new
+
+/--`getAllImports cmd id` takes a `Syntax` input `cmd` and returns the `NameSet` of all the
+module names that are implied by
+* the `SyntaxNodeKinds`,
+* the attributes of `cmd` (if there are any),
+* the identifiers contained in `cmd`,
+* if `cmd` adds a declaration `d` to the environment, then also all the module names implied by `d`.
+The argument `id` is optional and is expected to be the internally generated name of a "nameless"
+`instance`.
+-/
+def getAllImports (cmd : Syntax) (id : Syntax := default) (dbg? : Bool := false) :
+    CommandElabM NameSet := do
   let env ← getEnv
   --let nm ← liftCoreM do realizeGlobalConstNoOverloadWithInfo (getId cmd)
-  let ts := (getSyntaxNodeKinds cmd).append <| getVisited env (getId cmd).getId
+  -- we put together the implied declaration names, the `SyntaxNodeKinds`, the attributes
```

Comment:
```suggestion
  -- We collect the implied declaration names, the `SyntaxNodeKinds` and the attributes.
```
That said, reading this comment leaves me wondering "which attributes" or "that attributes of what"...

Reply:
These are the attributes of the declaration, e.g. `@[fun_prop] ...`.  Should I say more about this?

Reply:
I think this is okay.